### ISSUE TITLE
Import order matters between fiona and geopandas

### ISSUE THREAD

**Issue by spittssj** [geopandas/geopandas#1786]
- [X] I have read the [documentation on installation](https://geopandas.org/install.html) and followed the instructions provided.

- [X] I have looked through [issues labeled "installation"](https://github.com/geopandas/geopandas/labels/installation) in the geopandas repo.

- If your issue is related to installation using the `conda-forge` channel, please open an issue in [geopandas-feedstock repository](https://github.com/conda-forge/geopandas-feedstock) instead.

---

#### System information

MacOS Big Sur 15.1
Python 3.9.1 installed using homebrew.
With a fresh install of python@3.9 port, if I don't import fiona before geopandas, then I get a strange linking error below.

<details>
  <summary>Importing fiona before geopandas works perfectly</summary>
  <pre>

  (env) <username@machine> env % python
Python 3.9.1 (default, Jan  8 2021, 17:17:43) 
[Clang 12.0.0 (clang-1200.0.32.28)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import fiona
>>> import geopandas
>>> 

  </pre>
</details>

On the other hand,

<details>
  <summary>Importing geopandas by itself breaks</summary>
  <pre>
(env) stephenpitts@Ellacuria env % python
Python 3.9.1 (default, Jan  8 2021, 17:17:43) 
[Clang 12.0.0 (clang-1200.0.32.28)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import geopandas
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stephenpitts/Code/geo/env/lib/python3.9/site-packages/geopandas/__init__.py", line 7, in <module>
    from geopandas.io.file import _read_file as read_file  # noqa
  File "/Users/stephenpitts/Code/geo/env/lib/python3.9/site-packages/geopandas/io/file.py", line 7, in <module>
    import fiona
  File "/Users/stephenpitts/Code/geo/env/lib/python3.9/site-packages/fiona/__init__.py", line 86, in <module>
    from fiona.collection import BytesCollection, Collection
  File "/Users/stephenpitts/Code/geo/env/lib/python3.9/site-packages/fiona/collection.py", line 11, in <module>
    from fiona.ogrext import Iterator, ItemsIterator, KeysIterator
ImportError: dlopen(/Users/stephenpitts/Code/geo/env/lib/python3.9/site-packages/fiona/ogrext.cpython-39-darwin.so, 2): Symbol not found: _GEOSArea
  Referenced from: /usr/local/opt/libspatialite/lib/libspatialite.7.dylib
  Expected in: flat namespace
 in /usr/local/opt/libspatialite/lib/libspatialite.7.dylib
>>> 
  </pre>
</details>

<summary>Brew Package List</summary>
<details><pre>
autoconf 2.69
boost 1.75.0_1
cairo 1.16.0_4
cfitsio 3.490
cgal 5.2
eigen 3.3.9
epsilon 0.9.2
expat 2.2.10
file-formula 5.39_1
fontconfig 2.13.1
freetype 2.10.4
freexl 1.0.6
fribidi 1.0.10
gcc 10.2.0_2
gd 2.3.0
gdal 3.2.1
gdbm 1.18.1_1
gdk-pixbuf 2.42.2
geos 3.9.0
gettext 0.21
giflib 5.2.1
git 2.30.0
glib 2.66.4_1
gmp 6.2.1
gobject-introspection 1.66.1_1
graphite2 1.3.14
graphviz 2.44.1
gts 0.7.6_2
harfbuzz 2.7.4
hdf5 1.12.0_1
icu4c 67.1
isl 0.23
jasper 2.0.24
jpeg 9d
json-c 0.15
krb5 1.18.3
libdap 3.20.7
libffi 3.3_2
libgeotiff 1.6.0
libmagic 5.39
libmpc 1.2.1
libpng 1.6.37
libpq 13.1
librsvg 2.50.2
librttopo 1.1.0
libsodium 1.0.18_1
libspatialite 5.0.0
libtiff 4.2.0
libtool 2.4.6_2
libxml2 2.9.10_2
little-cms2 2.11
lzo 2.10
minizip 1.2.11
mpfr 4.1.0
netcdf 4.7.4_2
netpbm 10.86.18
node 15.5.1
nspr 4.29
nss 3.60.1
numpy 1.19.5
openblas 0.3.13
openjpeg 2.3.1 2.4.0
openssl@1.1 1.1.1i
pandoc 2.11.3.2
pango 1.48.0
pcre 8.44
pcre2 10.36
pixman 0.40.0
pkg-config 0.29.2_3
poppler 21.01.0
popt 1.18
postgresql 13.1
proj 7.2.1
protobuf 3.14.0
protobuf-c 1.3.3_3
pyenv 1.2.22
python@3.9 3.9.1_6
qt 5.15.2
readline 8.1
sfcgal 1.3.9_1
sqlite 3.34.0
swig 4.0.2
szip 2.1.1_1
tcl-tk 8.6.11
unixodbc 2.3.9
webp 1.1.0
xerces-c 3.2.3
xz 5.2.5
zeromq 4.3.3_1
zlib 1.2.11
zstd 1.4.8
</pre></details>

<summary>PIP Package Summary</summary>
<details><pre>
Package         Version
--------------- -----------
attrs           20.3.0
certifi         2020.12.5
click           7.1.2
click-plugins   1.1.1
cligj           0.7.1
Fiona           1.8.18
GDAL            3.2.1
geopandas       0.8.1
munch           2.5.0
numpy           1.19.5
pandas          1.2.1
pip             20.3.3
pyproj          3.0.0.post1
python-dateutil 2.8.1
pytz            2020.5
setuptools      51.1.1
Shapely         1.7.1
six             1.15.0
</pre>
</details>

<summary>Linker output from the library above</summary>
<details><pre>
(env) stephenpitts@Ellacuria env % otool -L /usr/local/opt/libspatialite/lib/libspatialite.7.dylib 
/usr/local/opt/libspatialite/lib/libspatialite.7.dylib:
	/usr/local/opt/libspatialite/lib/libspatialite.7.dylib (compatibility version 9.0.0, current version 9.1.0)
	/usr/local/opt/libtiff/lib/libtiff.5.dylib (compatibility version 12.0.0, current version 12.0.0)
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/local/opt/libxml2/lib/libxml2.2.dylib (compatibility version 12.0.0, current version 12.10.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
	/usr/local/opt/minizip/lib/libminizip.1.dylib (compatibility version 2.0.0, current version 2.0.0)
	/usr/local/opt/librttopo/lib/librttopo.1.dylib (compatibility version 3.0.0, current version 3.0.0)
	/usr/local/opt/freexl/lib/libfreexl.1.dylib (compatibility version 3.0.0, current version 3.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/local/opt/proj/lib/libproj.19.dylib (compatibility version 22.0.0, current version 22.1.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/local/opt/sqlite/lib/libsqlite3.0.dylib (compatibility version 9.0.0, current version 9.6.0)
	/usr/local/opt/geos/lib/libgeos_c.1.dylib (compatibility version 18.0.0, current version 18.2.0)
</pre></details>

## What you expected to happen
I expected geopandas to import fine above.

## Step-by-step reproduction instructions 
I have rebuilt the entire set of packages from source to no avail.

brew install -s python
brew install -s gdal
python -m venv env
cd env; source env/bin/activate
pip install --no-binary :all: geopandas

The setup worked under the previous release of MacOS X and homebrew, and my collaborator uses MacOS X and homebrew (but is afraid to upgrade after I shared my experience), so I am reluctant to switch to conda, as some forum posts I have seen have suggested.

I have read about subtle issues with import ordering and ctypes on various other posts about problems with Geopandas. I could attach "python -v" output from the above if someone were interested. I've used Python on and off for 10 years and I'm frustrated that I don't even know how to debug this one.

Thanks for any help.

Stephen

**Comment by jorisvandenbossche** [comment]
I suppose it is an order issue with importing `fiona` and `shapely` (could you try with those two, to check?), where pandas probably first imports shapely and then fiona.  
(fiona depends on GDAL which depends on spatialiate which depends on GEOS, and shapely also depends on GEOS. And there might be going something wrong with the GEOS being loaded, I *think*, because the error message mentions GEOSArea)

This combination is indeed known to give problems in certain situations. We are slowly fixing the ctypes usage in Shapely, but that's only for the future. 

Now, I am not familiar myself with how Howebrew works (I also don't have a Mac). It might also that the packages are not fully ready for the latest version of MacOSX (I know that many packages have been struggling with this). 

I do know that the conda-forge packages already work on the latest MacOSX, and that's what we generally recommend as a reliable install method. Now, that might be a big change for you, and for sure it should also work with homebrew.

**Comment by spittssj** [comment]
Joris,

Thanks for your detailed reply. Importing shapely works fine as well. I agree with your intuition: there is something happening on ctypes involving import order, but I don't know the answer. I rebuild libspatialite from source and sent the Homebrew people the log: nothing unusual. Is there any other information I could send you?

<pre>
stephenpitts@Ellacuria ~ % python3
Python 3.9.1 (default, Jan  8 2021, 17:17:43) 
[Clang 12.0.0 (clang-1200.0.32.28)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import shapely
>>> import fiona
>>> import geopandas
</pre>

I am moving from my former life as a Python web developer to my current life as a graduate student in Economics with more use for scientific computing Python libraries. I may bite the bullet and switch to conda-forge this weekend, as all roads are converging that way.

Do you know someone with a working Mac stack who could help me? 

Our web app and assorted database management scripts use geodjango + geopandas + Postgis and I have QGIS as well. Up until now I installed it all through Homebrew and used Python + PIP with virtual environments to manage it all. What is the recommended way to install it? 

I'm worried that the C dependencies (like GDAL) will get confused if I install some of the above through conda (python geodjango geopandas) and other parts through Homebrew (my Postgresql + Postgis works fine and I would like not to break it). 

Thanks for your help.

Stephen

**Comment by martinfleis** [comment]
I have macOS. The recommended way no matter the system is to use conda to manage environments and conda-forge channel to install packages. All you mentioned can be installed via conda from conda-forge, including QGIS, so my assumption is that it could solve all your issues in a relatively painless way.

> I'm worried that the C dependencies (like GDAL) will get confused if I install some of the above through conda (python geodjango geopandas) and other parts through Homebrew (my Postgresql + Postgis works fine and I would like not to break it)

That may happen, PostGIS depends on GEOS as well. Conda is able to manage environments in a way in which existing homebrew packages do not interfere with those installed in conda env.

**Comment by spittssj** [comment]
Martin, 

Thanks for your gentle prodding. This morning I completely pruned the Homebrew Python 3.9 and Homebrew of all GIS packages (to leave them as a system Python for things that have nothing to do with development), installed conda, and set up a new environment for my project.

1) Geopandas imports fine. So the script that generates our GIS fixtures runs successfully.

2) My GIS fixtures import into GeoDjango fine. 

3) I went ahead and set up Postgresql and Postgis within the environment, since I only use it for local development and testing. It's really slick. I can keep the same version of both with my collaborator and our hosting provider, and just run the server when I am actively developing. 

This GIST was very helpful, since PostgreSQL install is something I do once in a blue moon:
https://gist.github.com/gwangjinkim/f13bf596fefa7db7d31c22efd1627c7a

I think my resistance before came from an ingrained attitude from a decade ago that did not want to "waste" 780mb of disk space on an environment. But the disk space is worth the five hours I spent trying to debug this issue Thursday-Friday with no success. And this project is complex enough that it benefits from the strength of conda -- managing things outside of the pure Python environment. 

We are also starting to do some analysis scripts for the project that use Rmarkdown and I'm eager to let conda manage the R install as well.

4) I don't see a maintainer listed here: https://anaconda.org/conda-forge/postgis
Do you have any idea when they plan to upgrade to PostGIS 3.1 or how I can find out about that?  Ditto with PostgreSQL 13. 

For now the setup works well but I want to make sure it's actively maintained. 

I'm writing this up as much for the future Google record in case someone else is interested in the same things.

Thanks again for all of your help. I visited your website and my own interests are similar, except I hope to do machine learning on rural satellite data. This project is about a coffee cooperative and several hundred producer plots in Chiapas, Mexico. 
 
Stephen

**Comment by martinfleis** [comment]
> Do you have any idea when they plan to upgrade to PostGIS 3.1 or how I can find out about that? Ditto with PostgreSQL 13.

No clue. You can try to ping the relevant feedstock repository. https://github.com/conda-forge/postgis-feedstock https://github.com/conda-forge/postgresql-feedstock

**Comment by jorisvandenbossche** [comment]
Opening an issue might bring it to the maintainers attention. There is an automatic update (https://github.com/conda-forge/postgresql-feedstock/pull/82), but it seems that will need some human attention. 
(generally I think it *is* maintained, but the maintainers are managing *lots* of packages, so it might not always be quickly updated for new versions)

**Comment by adriantre** [comment]
I have the exact same error importing geopandas. My setup is gdal from homebrew, python from pyenv, geopandas using poetry (env manager for pip).

**Comment by spittssj** [comment]
I am still suspicious of the homebrew installs. The homebrew people were resistant to the idea that their package was causing the problem, but once I switched to the conda-forge packages, it worked. I have also installed geopandas 0.8.1 on MacPorts system-wide and it worked. 

Honestly, now that conda is managing my Python install, I have even less need for homebrew, and I think I might switch back to MacPorts. I found this link helpful to explain the reason why:
https://saagarjha.com/blog/2019/04/26/thoughts-on-macos-package-managers/

I wish the Homebrew people would add geopandas to their CI, if they want to be taken seriously as a platform for scientific computing.
